### PR TITLE
add proxy_fit_metric_map, speed up simulations by 4x

### PR DESCRIPTION
**Pull request by Innixma**
add proxy_fit_metric_map, speed up simulations by ~4x, cleanup fast_metric logic, add binary support to fast_log_loss

Added a trick to use `log_loss` metric during the fit of EnsembleSelection when eval_metric is `roc_auc`. `log_loss` is a good proxy metric for `roc_auc` while being much faster to compute, saving a lot of time in simulations.

Mainline:

```
Fitting Fold 1/5 (R1S1)...
    train_datasets: 104/104 | train_tasks: 1040
    test_datasets : 26/26 | test_tasks : 260
    n_configs=608/608
2023-05-03 07:52:31,321	INFO worker.py:1538 -- Started a local Ray instance.
1	: Train: 3.17 | 12.87s	| Tr Change: 2.841	| Conf Options: 608	| ray | CatBoost_r9_BAG_L1
2	: Train: 2.69 | 58.34s	| Tr Change: 0.471	| Conf Options: 607	| ray | NeuralNetFastAI_r35_BAG_L1
3	: Train: 2.4 | 79.4s	| Tr Change: 0.290	| Conf Options: 606	| ray | LightGBM_r12_BAG_L1
4	: Train: 2.28 | 100.5s	| Tr Change: 0.128	| Conf Options: 605	| ray | ExtraTrees_r23_BAG_L1
5	: Train: 2.17 | 120.45s	| Tr Change: 0.110	| Conf Options: 604	| ray | LightGBM_r146_BAG_L1
6	: Train: 2.09 | 141.09s	| Tr Change: 0.073	| Conf Options: 603	| ray | CatBoost_r12_BAG_L1
7	: Train: 2.05 | 161.86s	| Tr Change: 0.047	| Conf Options: 602	| ray | CatBoost_r72_BAG_L1
8	: Train: 2.02 | 181.43s	| Tr Change: 0.024	| Conf Options: 601	| ray | LightGBM_r82_BAG_L1
9	: Train: 2.0 | 201.27s	| Tr Change: 0.019	| Conf Options: 600	| ray | RandomForest_r16_BAG_L1
10	: Train: 1.99 | 221.73s	| Tr Change: 0.019	| Conf Options: 599	| ray | NeuralNetFastAI_r41_BAG_L1
selected ['CatBoost_r9_BAG_L1', 'NeuralNetFastAI_r35_BAG_L1', 'LightGBM_r12_BAG_L1', 'ExtraTrees_r23_BAG_L1', 'LightGBM_r146_BAG_L1', 'CatBoost_r12_BAG_L1', 'CatBoost_r72_BAG_L1', 'LightGBM_r82_BAG_L1', 'RandomForest_r16_BAG_L1', 'NeuralNetFastAI_r41_BAG_L1']
test_score: 2.3743008798881537
```

This PR (test score diff can generally be ignored, mostly random noise):

```
Fitting Fold 1/5 (R1S1)...
	train_datasets: 104/104 | train_tasks: 1040
	test_datasets : 26/26 | test_tasks : 260
	n_configs=608/608
2023-05-04 21:46:39,710	INFO worker.py:1538 -- Started a local Ray instance.
1	: Train: 3.17 | 11.78s	| Tr Change: 2.841	| Conf Options: 608	| ray | CatBoost_r9_BAG_L1
2	: Train: 2.66 | 26.57s	| Tr Change: 0.507	| Conf Options: 607	| ray | NeuralNetFastAI_r35_BAG_L1
3	: Train: 2.4 | 31.38s	| Tr Change: 0.261	| Conf Options: 606	| ray | LightGBM_r111_BAG_L1
4	: Train: 2.28 | 37.16s	| Tr Change: 0.113	| Conf Options: 605	| ray | ExtraTrees_r23_BAG_L1
5	: Train: 2.19 | 40.82s	| Tr Change: 0.090	| Conf Options: 604	| ray | LightGBM_r159_BAG_L1
6	: Train: 2.13 | 45.64s	| Tr Change: 0.067	| Conf Options: 603	| ray | NeuralNetFastAI_r34_BAG_L1
7	: Train: 2.06 | 51.32s	| Tr Change: 0.063	| Conf Options: 602	| ray | CatBoost_r62_BAG_L1
8	: Train: 2.02 | 56.73s	| Tr Change: 0.045	| Conf Options: 601	| ray | CatBoost_r90_BAG_L1
9	: Train: 1.98 | 59.43s	| Tr Change: 0.038	| Conf Options: 600	| ray | RandomForest_r28_BAG_L1
10	: Train: 1.95 | 64.03s	| Tr Change: 0.028	| Conf Options: 599	| ray | NeuralNetFastAI_r7_BAG_L1
selected ['CatBoost_r9_BAG_L1', 'NeuralNetFastAI_r35_BAG_L1', 'LightGBM_r111_BAG_L1', 'ExtraTrees_r23_BAG_L1', 'LightGBM_r159_BAG_L1', 'NeuralNetFastAI_r34_BAG_L1', 'CatBoost_r62_BAG_L1', 'CatBoost_r90_BAG_L1', 'RandomForest_r28_BAG_L1', 'NeuralNetFastAI_r7_BAG_L1']
test_score: 2.4499736042136506
```

### CODE REVIEW

File: autogluon_zeroshot/simulation/ensemble_selection_config_scorer.py [hunk 1]
```diff
@@ -34,10 +33,17 @@ def __init__(self,
                  max_fold: Optional[float] = None,
                  backend: str = 'native',
                  use_fast_metrics: bool = True,
+                 proxy_fit_metric_map: Optional[Union[dict, str]] = 'default',
```
Comment (geoalgo) [inline autogluon_zeroshot/simulation/ensemble_selection_config_scorer.py]: Can we rather do the following:
```suggestion
                 proxy_fit_metric_map: Optional[dict] = None,
```
and then
```
        if proxy_fit_metric_map is None:
            proxy_fit_metric_map = {'roc_auc': 'log_loss'}
```
as it is the standard way of dealing with defaults (otherwise it is very confusing, as passing None is also a default case)

File: autogluon_zeroshot/simulation/ensemble_selection_config_scorer.py [hunk 2]
```diff
@@ -34,10 +33,17 @@ def __init__(self,
...
                  ):
         """
         TODO: Add docstring
         :param use_fast_metrics: If True, will leverage optimized eval metrics to speed up config scoring.
+        :param proxy_fit_metric_map:
+            If eval_metric is among the keys in the `proxy_fit_metric_map` dictionary,
+            the value eval_metric will be used during the weighted ensemble fitting process as a proxy.
+            For example, the proxy metric could be faster to compute while producing a similar end result.
+            If None: Do not use proxy metrics, equivalent to {}.
+            If 'default': set to {'roc_auc': 'log_loss'}, making 'log_loss' a proxy to 'roc_auc'
```
Comment (geoalgo) [inline autogluon_zeroshot/simulation/ensemble_selection_config_scorer.py]: This is a bit confusing as we are passing a dict to pretend one metric is like the other one. Can we just allow to configure using `logloss`?

**Review by geoalgo — COMMENTED** [review]
Can we add test for those changes? Also I think it will be easier to allow to pick one metric or the other as it is effectively what we are doing.

File: autogluon_zeroshot/simulation/ensemble_selection_config_scorer.py (hunk 1 above)
Comment (Innixma) [reply autogluon_zeroshot/simulation/ensemble_selection_config_scorer.py]: Updated to None

File: autogluon_zeroshot/simulation/ensemble_selection_config_scorer.py (hunk 2 above)
Comment (Innixma) [reply autogluon_zeroshot/simulation/ensemble_selection_config_scorer.py]: We discussed offline on this, as configuring with logloss isn't the same result

**Comment by Innixma** [comment]
Updated so the default is None, meaning the proxy metric is not used unless explicitly set by the user.

**Comment by Innixma** [comment]
I've updated the code to respond to comments and added a TODO for a unit test.

I'm merging to simplify the amount of in-progress pull requests.
